--- ../../../alsa-kernel/core/seq/oss/seq_oss.c	2007-12-20 10:19:26.000000000 +0100
+++ seq_oss.c	2007-12-20 10:54:07.000000000 +0100
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  * OSS compatible sequencer driver
  *
@@ -21,6 +22,7 @@
  */
 
 #include <linux/init.h>
+#include <linux/smp_lock.h>
 #include <linux/moduleparam.h>
 #include <linux/mutex.h>
 #include <sound/core.h>
@@ -187,12 +189,25 @@
 	return snd_seq_oss_ioctl(dp, cmd, arg);
 }
 
-#ifdef CONFIG_COMPAT
+#if defined(CONFIG_COMPAT) && defined(CONFIG_SND_HAVE_NEW_IOCTL)
 #define odev_ioctl_compat	odev_ioctl
 #else
 #define odev_ioctl_compat	NULL
 #endif
 
+#ifndef CONFIG_SND_HAVE_NEW_IOCTL
+/* need to unlock BKL to allow preemption */
+static int odev_ioctl_old(struct inode *inode, struct file * file,
+			  unsigned int cmd, unsigned long arg)
+{
+	int err;
+	unlock_kernel();
+	err = odev_ioctl(file, cmd, arg);
+	lock_kernel();
+	return err;
+}
+#endif
+
 static unsigned int
 odev_poll(struct file *file, poll_table * wait)
 {
@@ -208,14 +223,20 @@
 
 static const struct file_operations seq_oss_f_ops =
 {
+#ifndef LINUX_2_2
 	.owner =	THIS_MODULE,
+#endif
 	.read =		odev_read,
 	.write =	odev_write,
 	.open =		odev_open,
 	.release =	odev_release,
 	.poll =		odev_poll,
+#ifdef CONFIG_SND_HAVE_NEW_IOCTL
 	.unlocked_ioctl =	odev_ioctl,
 	.compat_ioctl =	odev_ioctl_compat,
+#else
+	.ioctl =	odev_ioctl_old,
+#endif
 };
 
 static int __init
@@ -305,3 +326,5 @@
 	info_entry = NULL;
 }
 #endif /* CONFIG_PROC_FS */
+
+EXPORT_NO_SYMBOLS;
