--- ../alsa-kernel/drivers/portman2x4.c	2007-12-20 10:20:18.000000000 +0100
+++ portman2x4.c	2007-12-20 11:12:14.000000000 +0100
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  *   Driver for Midiman Portman2x4 parallel port midi interface
  *
@@ -610,7 +611,13 @@
 /*********************************************************************
  * parport stuff
  *********************************************************************/
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 24)
 static void snd_portman_interrupt(void *userdata)
+#elif defined(CONFIG_SND_NEW_IRQ_HANDLER)
+static void snd_portman_interrupt(int irq, void *userdata)
+#else
+static void snd_portman_interrupt(int irq, void *userdata, struct pt_regs *reg)
+#endif
 {
 	unsigned char midivalue = 0;
 	struct portman *pm = ((struct snd_card*)userdata)->private_data;
@@ -875,3 +882,5 @@
 
 module_init(snd_portman_module_init);
 module_exit(snd_portman_module_exit);
+
+EXPORT_NO_SYMBOLS;
