diff -urN a/eglibc-linaro-2.18-2013.10/bits/predefs.h b/eglibc-linaro-2.18-2013.10/bits/predefs.h
--- a/eglibc-linaro-2.18-2013.10/bits/predefs.h	2013-10-08 10:22:54.000000000 +0000
+++ b/eglibc-linaro-2.18-2013.10/bits/predefs.h	1970-01-01 00:00:00.000000000 +0000
@@ -1,30 +0,0 @@
-/* Copyright (C) 2005 Free Software Foundation, Inc.
-   This file is part of the GNU C Library.
-
-   The GNU C Library is free software; you can redistribute it and/or
-   modify it under the terms of the GNU Lesser General Public
-   License as published by the Free Software Foundation; either
-   version 2.1 of the License, or (at your option) any later version.
-
-   The GNU C Library is distributed in the hope that it will be useful,
-   but WITHOUT ANY WARRANTY; without even the implied warranty of
-   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-   Lesser General Public License for more details.
-
-   You should have received a copy of the GNU Lesser General Public
-   License along with the GNU C Library; if not, write to the Free
-   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
-   02111-1307 USA.  */
-
-#ifndef _PREDEFS_H
-#define _PREDEFS_H
-
-#ifndef _STDC_PREDEF_H
-# error "Never use <bits/predefs.h> directly; include <stdc-predef.h> instead."
-#endif
-
-/* We do support the IEC 559 math functionality, real and complex.  */
-#define __STDC_IEC_559__		1
-#define __STDC_IEC_559_COMPLEX__	1
-
-#endif /* predefs.h */
diff -urN a/eglibc-linaro-2.18-2013.10/ChangeLog.eglibc b/eglibc-linaro-2.18-2013.10/ChangeLog.eglibc
--- a/eglibc-linaro-2.18-2013.10/ChangeLog.eglibc	2013-10-08 10:22:54.000000000 +0000
+++ b/eglibc-linaro-2.18-2013.10/ChangeLog.eglibc	2015-09-23 19:09:59.766877538 +0000
@@ -1,3 +1,11 @@
+2015-09-23  Dragan Stancevic  <dragans@broadcom.com>
+
+	* include/stdc-predef.h:  Added predefs.h contents.
+	* include/stdc-predef.h:  Added if __arm__ macros.
+	* Makefile (headers): Remove bits/predefs.h.
+	* bits/predefs.h: Remove file.
+	* ports/sysdeps/arm/bits/predefs.h: Remove file.
+
 2013-05-17  Joseph Myers  <joseph@codesourcery.com>
 
 	* configure.in, math/Makefile, math/gen-libm-test.pl,
diff -urN a/eglibc-linaro-2.18-2013.10/include/stdc-predef.h b/eglibc-linaro-2.18-2013.10/include/stdc-predef.h
--- a/eglibc-linaro-2.18-2013.10/include/stdc-predef.h	2013-10-08 10:22:54.000000000 +0000
+++ b/eglibc-linaro-2.18-2013.10/include/stdc-predef.h	2015-09-23 18:56:28.815435933 +0000
@@ -26,8 +26,20 @@
    explicitly includes a system header.  GCC knows the name of this
    header in order to preinclude it.  */
 
-/* Define __STDC_IEC_559__ and other similar macros.  */
-#include <bits/predefs.h>
+#if defined(__arm__) // __arm__
+/* We do support the IEC 559 math functionality, real and complex, but only
+   if a VFP coprocessor is present. If we don't have one, we fall back to
+   software emulation and the functions won't work properly. So in general,
+   we don't claim to support this functionality.  */
+#if defined (__VFP_FP__) && !defined(__SOFTFP__)
+#define __STDC_IEC_559__		1
+#define __STDC_IEC_559_COMPLEX__	1
+#endif
+#else // __arm__
+/* We do support the IEC 559 math functionality, real and complex.  */
+#define __STDC_IEC_559__		1
+#define __STDC_IEC_559_COMPLEX__	1
+#endif // __arm__
 
 /* wchar_t uses ISO/IEC 10646 (2nd ed., published 2011-03-15) /
    Unicode 6.0.  */
@@ -37,3 +49,4 @@
 #define __STDC_NO_THREADS__		1
 
 #endif
+
diff -urN a/eglibc-linaro-2.18-2013.10/Makefile b/eglibc-linaro-2.18-2013.10/Makefile
--- a/eglibc-linaro-2.18-2013.10/Makefile	2013-10-08 10:22:54.000000000 +0000
+++ b/eglibc-linaro-2.18-2013.10/Makefile	2015-09-23 18:46:43.875537704 +0000
@@ -59,7 +59,7 @@
 		   $(addprefix install-, no-libc.a bin lib data headers others)
 
 headers := limits.h values.h features.h gnu-versions.h bits/libc-lock.h \
-	   bits/xopen_lim.h gnu/libc-version.h stdc-predef.h bits/predefs.h
+	   bits/xopen_lim.h gnu/libc-version.h stdc-predef.h
 
 echo-headers: subdir_echo-headers
 
diff -urN a/eglibc-linaro-2.18-2013.10/ports/sysdeps/arm/bits/predefs.h b/eglibc-linaro-2.18-2013.10/ports/sysdeps/arm/bits/predefs.h
--- a/eglibc-linaro-2.18-2013.10/ports/sysdeps/arm/bits/predefs.h	2013-10-08 10:22:54.000000000 +0000
+++ b/eglibc-linaro-2.18-2013.10/ports/sysdeps/arm/bits/predefs.h	1970-01-01 00:00:00.000000000 +0000
@@ -1,35 +0,0 @@
-/* Copyright (C) 2005, 2006 Free Software Foundation, Inc.
-   This file is part of the GNU C Library.
-
-   The GNU C Library is free software; you can redistribute it and/or
-   modify it under the terms of the GNU Lesser General Public
-   License as published by the Free Software Foundation; either
-   version 2.1 of the License, or (at your option) any later version.
-
-   The GNU C Library is distributed in the hope that it will be useful,
-   but WITHOUT ANY WARRANTY; without even the implied warranty of
-   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-   Lesser General Public License for more details.
-
-   You should have received a copy of the GNU Lesser General Public
-   License along with the GNU C Library; if not, write to the Free
-   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
-   02111-1307 USA.  */
-
-#ifndef _PREDEFS_H
-#define _PREDEFS_H
-
-#ifndef _STDC_PREDEF_H
-# error "Never use <bits/predefs.h> directly; include <stdc-predef.h> instead."
-#endif
-
-/* We do support the IEC 559 math functionality, real and complex, but only
-   if a VFP coprocessor is present. If we don't have one, we fall back to
-   software emulation and the functions won't work properly. So in general,
-   we don't claim to support this functionality.  */
-#if defined (__VFP_FP__) && !defined(__SOFTFP__)
-#define __STDC_IEC_559__		1
-#define __STDC_IEC_559_COMPLEX__	1
-#endif
-
-#endif /* predefs.h */
