Upstream-status: backport

From 28d708c44bc47b56f6551ff285f78edcf61c208a Mon Sep 17 00:00:00 2001
From: Marc-Antoine Perennou <Marc-Antoine@Perennou.com>
Date: Thu, 31 Oct 2013 12:37:50 +1000
Subject: [PATCH] Accept make versions 4.0 and greater

---
 ChangeLog    | 6 ++++++
 NEWS         | 2 +-
 configure    | 2 +-
 configure.ac | 2 +-
 4 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/configure b/configure
index f382138..5e61abd 100755
--- a/configure
+++ b/configure
@@ -4761,7 +4761,7 @@ $as_echo_n "checking version of $MAKE... " >&6; }
   ac_prog_version=`$MAKE --version 2>&1 | sed -n 's/^.*GNU Make[^0-9]*\([0-9][0-9.]*\).*$/\1/p'`
   case $ac_prog_version in
     '') ac_prog_version="v. ?.??, bad"; ac_verc_fail=yes;;
-    3.79* | 3.[89]*)
+    3.79* | 3.[89]* | [4-9].* | [1-9][0-9]*)
        ac_prog_version="$ac_prog_version, ok"; ac_verc_fail=no;;
     *) ac_prog_version="$ac_prog_version, bad"; ac_verc_fail=yes;;
 
-- 
1.9.4

