diff --git media/filters/stream_parser_factory.cc media/filters/stream_parser_factory.cc
index 4a36875..daec99c 100644
--- media/filters/stream_parser_factory.cc
+++ media/filters/stream_parser_factory.cc
@@ -333,38 +333,14 @@ static bool VerifyCodec(
     std::vector<CodecInfo::HistogramTag>* video_codecs) {
   switch (codec_info->type) {
     case CodecInfo::AUDIO:
+      if (codec_info->tag == CodecInfo::HISTOGRAM_OPUS)
+        return false;
       if (audio_codecs)
         audio_codecs->push_back(codec_info->tag);
       return true;
     case CodecInfo::VIDEO:
-#if defined(OS_ANDROID)
-      // TODO(wolenetz, dalecurtis): This should instead use MimeUtil() to avoid
-      // duplication of subtle Android behavior.  http://crbug.com/587303.
-      if (codec_info->tag == CodecInfo::HISTOGRAM_H264) {
-        if (media::IsUnifiedMediaPipelineEnabled() &&
-            !media::HasPlatformDecoderSupport()) {
-          return false;
-        }
-
-        if (!MediaCodecUtil::IsMediaCodecAvailable())
-          return false;
-      }
-      if (codec_info->tag == CodecInfo::HISTOGRAM_VP8 &&
-          !media::MediaCodecUtil::IsVp8DecoderAvailable() &&
-          !media::IsUnifiedMediaPipelineEnabled()) {
+      if (codec_info->tag == CodecInfo::HISTOGRAM_VP9)
         return false;
-      }
-      if (codec_info->tag == CodecInfo::HISTOGRAM_VP9 &&
-          !media::MediaCodecUtil::IsVp9DecoderAvailable() &&
-          !media::IsUnifiedMediaPipelineEnabled()) {
-        return false;
-      }
-      if (codec_info->tag == CodecInfo::HISTOGRAM_OPUS &&
-          !media::PlatformHasOpusSupport() &&
-          !media::IsUnifiedMediaPipelineEnabled()) {
-        return false;
-      }
-#endif
       if (video_codecs)
         video_codecs->push_back(codec_info->tag);
       return true;
